tests/integration/test_healing.py

```diff
@@ -0,0 +1,298 @@
+            lambda: len(ops_test.model.applications[APP_NAME].units) == 3
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME],
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+        )
+
+
+@pytest.mark.order(1)
+@pytest.mark.abort_on_fail
+@pytest.mark.healing_tests
+async def test_kill_db_process(ops_test: OpsTest) -> None:
+    """Kill mysqld process and check for auto cluster recovery."""
+    await build_and_deploy(ops_test)
+
+    app = await app_name(ops_test)
+
+    primary_unit = await get_primary_unit_wrapper(ops_test, app)
+    another_unit = (set(ops_test.model.applications[app].units) - {primary_unit}).pop()
+
+    primary_unit_ip = await get_unit_ip(ops_test, primary_unit.name)
+
+    # get running mysqld PID
+    pid = await get_process_pid(ops_test, primary_unit.name, MYSQL_DAEMON)
+
+    # kill mysqld for the unit
+    logger.info(f"Killing process id {pid}")
+    await ops_test.juju("ssh", primary_unit.name, "sudo", "kill", "-9", pid)
+
+    config = {
+        "username": CLUSTER_ADMIN_USERNAME,
+        "password": await get_system_user_password(primary_unit, CLUSTER_ADMIN_USERNAME),
+        "host": primary_unit_ip,
+    }
+
+    # retrieve new PID
+    new_pid = await get_process_pid(ops_test, primary_unit.name, MYSQL_DAEMON)
+    logger.info(f"New process id is {new_pid}")
+
+    # verify that mysqld instance is not the killed one
+    assert new_pid != pid, "❌ PID for mysql daemon did not changed"
+
+    # verify daemon restarted via connection
+    assert is_connection_possible(config), f"❌ Daemon did not restart on unit {primary_unit.name}"
+
+    # verify instance is part of the cluster
+    logger.info("Check if instance back in cluster")
+    assert await is_unit_in_cluster(
+        ops_test, primary_unit.name, another_unit.name
+    ), " Unit not online in the cluster"
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.healing_tests
+async def test_freeze_db_process(ops_test: OpsTest):
+    """Freeze and unfreeze process and check for auto cluster recovery."""
+    app = await app_name(ops_test)
+    primary_unit = await get_primary_unit_wrapper(ops_test, app)
+    another_unit = (set(ops_test.model.applications[app].units) - {primary_unit}).pop()
+
+    primary_unit_ip = await get_unit_ip(ops_test, primary_unit.name)
+
+    # get running mysqld PID
+    pid = await get_process_pid(ops_test, primary_unit.name, MYSQL_DAEMON)
+
+    # freeze (STOP signal) mysqld for the unit
+    logger.info(f"Freezing process id {pid}")
+    await ops_test.juju("ssh", primary_unit.name, "sudo", "kill", "-19", pid)
+
+    config = {
+        "username": CLUSTER_ADMIN_USERNAME,
+        "password": await get_system_user_password(primary_unit, CLUSTER_ADMIN_USERNAME),
+        "host": primary_unit_ip,
+    }
+
+    # verify that connection is not possible
+    assert not is_connection_possible(config), "❌ Mysqld is not paused"
```

Comment:
should we be checking if a database unit (not the primary) is still reachable when the primary unit's mysqld is paused?

Reply:
why not? I'll add it

Reply:
Will add on the follow up PR